### DISCUSSION TITLE
tooltips 自定义用法

### DISCUSSION THREAD

**Discussion by WeiShengv99** [antvis/S2#849] [Q&A] [answered]
当我想要自定义tooltips的时候，我发现我并不能根据数据去做一些逻辑判断

例如
![image](https://user-images.githubusercontent.com/35689628/144186919-8ba40b91-aa41-4100-af59-79db88767042.png)

如果我点击 杭州这行的时候， 我期望根据 `杭州` 来判断  展示一些关于 这一行的信息的描述

目前已有的api 里  ` row  content` 只能定义所有的 行头 ,
 ```
getSpreadSheet={(instance) => {instance.showTooltip = (tooltipOptions) => {}
``` 
修改的方式只能获取到当前选择的 单元格数据, summaries 字段看起来不适合做这种自定义判断
![image](https://user-images.githubusercontent.com/35689628/144187356-e748bb9a-8d85-4e03-ac0c-e8cc8704da1d.png)

关于 行和列头的自定义 tooltips ，怎么做才是最佳实践？期待你的回复

**Comment by WeiShengv99** [comment]
对于 hover 行头需要提供行提示的需求，我目前是这么去实现的，这样基本上可以解决我的需求

如果    `this.spreadsheet.tooltip.show` 的时候 我可以传递一个 自定义的 `className `，可以让我在使用默认 content 组件的时候去改变一些样式上的设置，这样我就不会丢失 默认的 content组件的行为。


![image](https://user-images.githubusercontent.com/35689628/144198264-ad1fa5be-002f-4052-9102-6ad41299ec17.png)

**Comment by lijinke666** [comment]
通过自定义交互 确实可以实现你的需求, 但感觉对用户来说成本有点太大了, 自定义这一块我们确实还有一些欠缺, 内部讨论中...

似乎 `options.tooltip.content` 支持回调和直接渲染两种方式 能最大程度满足不同用户的自定义需求

```ts
const s2Options = {
  tooltip: {
     // cell: 当前触发的tooltip的单元格, const meta = cell.getMeta() 之类的
     content: (cell) => <YourComponent {...meta} />
     // 或者
     content: <YourComponent /> 
  }
}
```

**Reply by WeiShengv99** [reply]
@lijinke666 
这个解决方案有点瑕疵，而且只会报一次错误
```
this.spreadsheet.tooltip.show({
        position: { x: event.clientX, y: event.clientY },
        content: (
          <div
            className={Style.row_cell}
            key={`${item?.id}${item?.description}`}
          >
            <div>{item?.dimensionName}</div>
            <div>{item?.description}</div>
          </div>
        ),
      });
```
![image](https://user-images.githubusercontent.com/35689628/144206568-e846f72d-55fc-42ad-9562-3a866a2de943.png)


这里是复现的 [demo](https://codesandbox.io/s/hardcore-booth-q26q1?file=/src/App.tsx)
大概原因是因为   自定义组件里有2个 子元素 就会有key 问题

**Comment by lijinke666** [comment, marked as answer]
@WeiShengv99 感觉你是仔细看了我们的文档的 👍🏻 我们还有很多可以改进的地方, 请问目前你使用 S2 的整体感觉如何呢? 欢迎更多像你这样的用户给我们提出更多宝贵意见

**Reply by WeiShengv99** [reply]
我是因为遇到透视表格（并且是大数据的情况下），才找到这个库的。
目前社区里能支持这个需求的库基本没有比较好的解决方案，最近我才发现 s2 可以基本解决我的需求。👍🏻

整理来看体验是不错的，功能很强大👍🏻

目前在使用过程中遇到的问题一般都是在  查阅文档的时候产生的困惑
比如 

 BaseTooltip 和 BaseDataSet 这些是 404 的页面， 对于 dataSet 的描述 好像没有看见（应该是类似 g2的 数据集的概念吧）

![image](https://user-images.githubusercontent.com/35689628/144202394-b53ecc70-b3ee-4d3d-a1bc-ec0621f62ae4.png)


 getMeta 等api 返回的信息也没有相关说明。我是通过 console 和 自定义交互 的demo 才了解到, 文档站的api 检索功能比较薄弱

![image](https://user-images.githubusercontent.com/35689628/144204992-a80ff282-4c2e-4e4d-9f02-727a0fd9a2ff.png)

**Reply by xingwanying** [reply]
您好，首先非常感谢您的认可和反馈，文档部分的反馈已收到，我们会不断完善。
关于tooltip的自定义目前有结合点击事件hook的方式来实现，demo可以参考下
https://s2.antv.vision/zh/examples/react-component/tooltip#custom-interaction-tooltip

![image](https://user-images.githubusercontent.com/10885578/144207696-6b5ab8a7-94d1-44a9-beea-93c203b2f33d.png)

hover 的实现也会是类似的思路，但目前还未暴露出 hover 事件的hook。您看下这样的方式是否能较为简单地满足需求？

**Reply by xingwanying** [reply]
另外，您有钉钉号方便留一下么？我们有内部交流群可以一对一快速响应您的问题～

**Reply by WeiShengv99** [reply]
没有相关的  钉钉 ，后续有比较急的问题 我再下载一个吧
onRowCellClick 是一个不错的解决方案，谢谢你的建议，我大概知道怎么做了。
